core/services/vrf/proof_response.go

```diff
@@ -0,0 +1,109 @@
+package vrf
+
+// Contains logic/data for mandatorily mixing VRF seeds with the hash of the
+// block in which a VRF request appeared
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+// FinalSeed(preSeed, blockHash) is the seed which is actually passed to the VRF
+// proof generator, given the preseed and the hash of the block in which the
+// VRFCoordinator emitted the log for the request this is responding to.
+func FinalSeed(preSeed Seed, blockHash common.Hash) (finalSeed *big.Int) {
```

Comment:
Do we trust all of these inputs?

Reply:
Could you say a bit more about the trust parameters you have in  mind?

The intention is that the `preSeed` and the `blockHash` are [read out of the randomness-request ethereum-log which triggers the `randomness` initiator](https://github.com/smartcontractkit/chainlink/blob/172899821-vrf-blockhash-included-in-final-seed/core/store/models/parse_randomness_request.go#L18). So they are trusted to the same extent as the ethereum node to which the chainlink node is connected. The `preSeed` is generated deterministically from input provided by the consuming contract, and the `blockHash` comes from the block which generated the randomness-request log. So they are both controllable by untrusted parties, to a certain extent: the seed by the requester, the hash by the miner (at fairly high cost.)

The `preSeed` is morally just a `uint256`, and is passed through code which is intended to be able to handle any `uint256` value without a hiccup.

The output from `FinalSeed` is a hash of those two inputs, and is again morally just a `uint256`, and intended to be handled by code which should handle any `uint256`.

Reply:
Cool, just want to be aware for example that it's not possible for an attacker to generate a malformed log message that causes one of these binaries to have zero length or something.